src/compiler/sys.rs

```diff
@@ -0,0 +1,45 @@
+use std::{env, fs};
+
+pub trait System {
+    fn args(&self) -> &Vec<String>;
+    fn read_file(&self, path: &str) -> Option<String>;
+    fn get_current_directory(&self) -> String;
+}
+
+struct SystemConcrete {
+    args: Vec<String>,
+}
+
+impl SystemConcrete {
+    fn read_file_worker(&self, file_name: &str) -> Option<String> {
+        fs::read_to_string(file_name).ok()
+    }
+}
+
+impl System for SystemConcrete {
+    fn args(&self) -> &Vec<String> {
+        &self.args
+    }
+
+    fn read_file(&self, file_name: &str) -> Option<String> {
+        self.read_file_worker(file_name)
+    }
+
+    fn get_current_directory(&self) -> String {
+        env::current_dir()
+            .unwrap()
+            .into_os_string()
+            .into_string()
+            .unwrap()
+    }
+}
+
+lazy_static! {
+    static ref SYS: SystemConcrete = SystemConcrete {
+        args: env::args().skip(1).collect(),
+    };
+}
+
+pub fn get_sys() -> &'static impl System {
+    &*SYS
+}
```

Comment:
What does this get you over directly exposing `SYS` from the `lazy_static`?

Reply:
I ran into the fact that I guess `lazy_static!` actually [creates a one-off type under the hood](https://stackoverflow.com/a/48115258/732366) for the static variable which then meant that I couldn't type `SYS` as implementing `System` (because the one-off type doesn't "inherit" the trait implementations of the declared type I guess)